drivers: i2c: Base OMAP I2C support for TI-K3 series processor

malto101 opened the pull request:
This PR adds the support for I2C serial interface on TI K3 series. It is compatible with Philips I2C physical layer. 
The commit includes:
Zephyr i2c api implementation
speed Select(100Kbps and 400Kbps)
Bus recovery

This driver is tested on BeagleBone AI64
Note: This PR is based [#71527](https://github.com/zephyrproject-rtos/zephyr/pull/71527)

malto101 committed:
drivers: i2c: Base OMAP I2C support for TI-K3 processor

The OMAP I2C provides support for I2C serial interface on TI K3 series.
It is compatible with Philips I2C physical layer.
The commit includes:
Zephyr i2c api implementation
Polling Mode

Signed-off-by: Dhruv Menon <dhruvmenon1104@gmail.com>

henrikbrixandersen commented on drivers/i2c/Kconfig.omap:
```diff
@@ -0,0 +1,25 @@
+
+config I2C_OMAP
+	bool "TI OMAP I2C Driver"
+	default y
```
```suggestion
	default y
	depends on DT_HAS_TI_OMAP_I2C_ENABLED
```

malto101 replied:
Added the suggestion.

nordicjm commented on drivers/i2c/Kconfig.omap:
```diff
@@ -0,0 +1,25 @@
+	help
+		Enable the I2C driver for TI OMAP SoCs.
+
+menu "Custom I2C OMAP Logging"
+config OMAP_I2C_LOG_LEVEL
+    int "OMAP I2C Log Level"
+    default 4  # For DEBUG level
+    range 0 4
+    help
+      Set the log level for the OMAP I2C module:
+        0: None
+        1: Error
+        2: Warning
+        3: Info
+        4: Debug
```
Remove this and use the existing i2c log level

vaishnavachath commented on dts/arm/ti/j721e_main_r5.dtsi:
```diff
@@ -0,0 +1,107 @@
+		interrupt-parent = <&vim>;
+		status = "disabled";
+	};
+};
```
Can you drop these J72E R5 support addition changes from this PR? there is no direct dependency between that PR/these changes and the driver addition

malto101 replied:
Understood!
I will add it as a separate PR

vaishnavachath commented on drivers/i2c/i2c_omap.c:
```diff
@@ -0,0 +1,947 @@
+	static void i2c_omap_##inst##_init(const struct device *dev)                               \
+	{                                                                                          \
+		IRQ_CONNECT(DT_INST_IRQN(inst), 0, i2c_omap_isr, DEVICE_DT_INST_GET(inst), 0);     \
+                                                                                                   \
```
you will need to honor IRQ flags here, not hardcode as 0 always.

malto101 replied:
Updated!

vaishnavachath replied:
Looks like this change was missed? Also do you really need to enable the irq during init given that you only support polled mode?

malto101 replied:
Not really,
Will drop it for now and add it with IRQ support

vaishnavachath commented on drivers/i2c/i2c_omap.c:
```diff
@@ -0,0 +1,947 @@
+	i2c_omap_reg_set(cfg, I2C_SYSTEST, reg);
+
+	i2c_bitbang_init(&bitbang_omap, &bitbang_omap_io, (void *)cfg);
+
```
Can you explain what is being done here? who cleans up after recovery? can you explain in the commit message the details of the implementation of the driver atleast wherever it deviates from standard implementations like Linux/U-Boot/TI SDKs?

malto101 replied:
I based out the recovery from [i2c_bitbang_recover_bus](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/i2c/i2c_bitbang.c#L277), which enables bit bang operations to send 9 SCL clk pulses.

Post that, it checks if the line is down. and then disables the system_test register and . resuming the transaction 

Will update this in the commit message

malto101 replied:
Done!

vaishnavachath commented on drivers/i2c/i2c_omap.c:
```diff
@@ -0,0 +1,947 @@
+typedef void (*init_func_t)(const struct device *dev);
+
+struct i2c_omap_cfg {
+	uint32_t base;
```
This is not going to scale well to any other platform [or any platform currently in upstream] using this driver as there is a RAT/MMU for all other platforms you will need to use DEVICE_MMIO_* API to make it work for other platforms. See https://docs.zephyrproject.org/apidoc/latest/group__device-mmio.html

malto101 replied:
Updated it to use DEVICE_MMIO_*
Thank you for the reference!

vaishnavachath commented on drivers/i2c/i2c_omap.c:
```diff
@@ -0,0 +1,947 @@
+#define I2C_BUF_RXFIF_CLR BIT(14) /* RX FIFO Clear */
+#define I2C_BUF_XDMA_EN   BIT(7)  /* TX DMA channel enable */
+#define I2C_BUF_TXFIF_CLR BIT(6)  /* TX FIFO Clear */
+
```
Add these back when you actually add these features like DMA support?

vaishnavachath commented on drivers/i2c/i2c_omap.c:
```diff
@@ -0,0 +1,947 @@
+#define I2C_DMARXENABLE_CLR 0x40 /* Per-event DMA RX enable clear */
+#define I2C_DMATXENABLE_CLR 0x44 /* Per-event DMA TX enable clear */
+#define I2C_DMARXWAKE_EN    0x48 /* DMA RX wake enable */
+#define I2C_DMATXWAKE_EN    0x4c /* DMA TX wake enable */
```
Lot of unused defines? is this copied from some existing TI driver? then you need to retain the original copyright as well.

vaishnavachath commented on drivers/i2c/i2c_omap.c:
```diff
@@ -0,0 +1,921 @@
+static inline void i2c_omap_reg_set(const struct i2c_omap_cfg *cfg, uint32_t reg, uint32_t value)
+{
+	sys_write16(value, cfg->base + reg);
+}
```
These are all 32-bit registers as per the TRM, even though the upper bits are reserved , that does not mean you should access it as 16-bit registers?

malto101 replied:
Understood!
I will change it to sys32.

vaishnavachath commented on drivers/i2c/i2c_omap.c:
```diff
@@ -0,0 +1,921 @@
+	/* Take the I2C module out of reset */
+	i2c_omap_reg_set(cfg, I2C_CON, 0);
+	i2c_omap_set_speed(dev, speed_cfg);
+	i2c_omap_reset(dev);
```
what is the need for resetting the module when initializing/taking the module out of reset? I don't see corresponding reset in reference Linux driver.

malto101 replied:
Removed

malto101 commented on drivers/i2c/i2c_omap.c:
```diff
@@ -0,0 +1,921 @@
+
+/* I2C Registers */
+#define I2C_SYSC          0x10 /* System configuration */
+#define I2C_IRQENABLE_SET 0x2c /* Per-event interrupt enable bit set */
```
Got it!
I will change it accordingly.

malto101 commented on drivers/i2c/i2c_omap.c:
```diff
@@ -0,0 +1,921 @@
+
+	/* If configured for High Speed */
+	switch (speed) {
+	case I2C_BITRATE_FAST + 1 ... UINT32_MAX:
```
I plan to add support to super fast mode plus(1 Mbps) too, once this PR gets merged

Would it be better if I remove this, and when adding for 1 Mbps, bring it back?

malto101 commented on drivers/i2c/i2c_omap.c:
```diff
@@ -0,0 +1,921 @@
+ */
+static int i2c_transfer_message(const struct device *dev, struct i2c_msg *msg, bool polling,
+				uint16_t addr)
+{
```
It is not a generic function. The function handles the message transfer by setting the various config.
Will change it to be more specific.

malto101 commented on drivers/i2c/i2c_omap.c:
```diff
@@ -0,0 +1,921 @@
+ */
+static int __maybe_unused i2c_omap_transfer_irq(const struct device *dev, struct i2c_msg msgs[],
+						uint8_t num_msgs, uint16_t addr)
+{
```
At the moment, it is not being used. I will remove it and add as a separate PR

malto101 commented on drivers/i2c/i2c_omap.c:
```diff
@@ -0,0 +1,921 @@
+
+	/* Wait for the transfer to complete, either by polling or using a semaphore */
+	if (!polling) {
+		result = k_sem_take(&data->cmd_complete, K_MSEC(I2C_OMAP_TIMEOUT));
```
IRQ mode is not currently fully supported.
I will remove it and add it as a new PR

malto101 committed:
drivers: i2c: Added Bus recovery support to OMAP I2C.

The bus recovery feature utilizing bit-bang operations, which sends
SCL clock pulses to recover the bus and checks if the line is down.
Upon recovery, it disables the system test register before resuming
the transactions.

Signed-off-by: Dhruv Menon <dhruvmenon1104@gmail.com>

malto101 commented on drivers/i2c/i2c_omap.c:
```diff
@@ -0,0 +1,891 @@
+ * @param reg The register address to read from.
+ * @return The value read from the register.
+ */
+static inline uint16_t i2c_omap_reg_access(const struct i2c_omap_cfg *cfg, int reg)
```
The upper bits of the register were reserved. I will update the return type from uint16_t to uint32_t for consistency with sys_read32

malto101 commented on drivers/i2c/i2c_omap.c:
```diff
@@ -0,0 +1,891 @@
+};
+
+#define I2C_OMAP_INIT(inst)                                                                        \
+	LOG_INSTANCE_REGISTER(omap_i2c, inst, 4);                                                  \
```
I will change it!

malto101 commented on drivers/i2c/i2c_omap.c:
```diff
@@ -0,0 +1,891 @@
+	switch (I2C_SPEED_GET(dev_config)) {
+	case I2C_SPEED_STANDARD:
+		speed_cfg = I2C_BITRATE_STANDARD;
+		goto out;
```
Understood!
Will change it.

malto101 commented on drivers/i2c/i2c_omap.c:
```diff
@@ -0,0 +1,891 @@
+
+	buf_cfg = i2c_omap_reg_access(cfg, I2C_OMAP_BUF);
+	if (data->receiver) {
+		buf_cfg &= ~(0x3f << 8);
```
As per the TRM document ([J721E_registers4.pdf, Page 149](https://www.ti.com/lit/zip/spruil1)), there are Rx and Tx FIFO clear bits that we don’t want to overwrite during this function call.

malto101 replied:
Removed it as some config bits were not required
Thank you for feedback @dnltz !

malto101 commented on drivers/i2c/i2c_omap.c:
```diff
@@ -0,0 +1,891 @@
+
+	i2c_omap_reg_set(cfg, I2C_OMAP_CON, I2C_OMAP_CON_EN);
+
+	k_busy_wait(1);
```
Removed it!

dnltz commented on drivers/i2c/i2c_omap.c:
```diff
@@ -0,0 +1,820 @@
+
+typedef void (*init_func_t)(const struct device *dev);
+#define DEV_I2C_BASE(dev) ((I2C_OMAP_Type *)DEVICE_MMIO_NAMED_GET(dev, base))
+
```
You could consider to add more a define for data and config. See https://github.com/zephyrproject-rtos/zephyr/blob/31fea97e05fd524b089c70281356d07df7788349/drivers/gpio/gpio_davinci.c#L28

It's not a huge improvement but in case a struct changes, this has to be changed only at one location.

malto101 replied:
This makes sense.
Thanks for showing this.
Will update it!

dnltz commented on drivers/i2c/i2c_omap.c:
```diff
@@ -0,0 +1,820 @@
+	__IO uint32_t SCLH;          /**< SCL High Time, offset: 0xB8 */
+	__IO uint32_t SYSTEST;       /**< System Test, offset: 0xBC */
+	__IO uint32_t BUFSTAT;       /**< Buffer Status, offset: 0xC0 */
+} I2C_OMAP_Type, *I2C_OMAP_Type_Ptr;
```
Upper-case is not common for typedefs. Additional, don't create one for a pointer. It's literally shorter to add the star character in front instead of typing `_ptr`.

```suggestion
} i2c_omap_regs_t;
```

malto101 replied:
Updated!

dnltz commented on drivers/i2c/i2c_omap.c:
```diff
@@ -0,0 +1,820 @@
+ * @param dev Pointer to the device structure for the I2C controller.
+ * @return 0 if successful, negative error code otherwise.
+ */
+static int i2c_omap_init(const struct device *dev)
```
It's common to have the init function at the end before `I2C_OMAP_INIT` .

malto101 replied:
Got it!
Will update this.

dnltz reviewed (commented):
Found some nitpicks. I can test this driver next week on our AM64 board.

malto101 commented:
Thank you very much for your review @dnltz! I would appreciate it if you could test this on the AM64, as I’ve only had the chance to test it on the BBAI64. Please let me know if there’s any additional room for improvement.

Thank you again!

glneo commented on drivers/i2c/i2c_omap.c:
```diff
@@ -0,0 +1,815 @@
+ * @param dev Pointer to the device structure.
+ * @param stat The status value to be written to the I2C_OMAP_STAT register.
+ */
+static inline void i2c_omap_ack_stat(const struct device *dev, uint16_t stat)
```
Everywhere this function is used, `i2c_base_addr` is already known, you could just pass that into this function.

Or even better, since that would make this function a one liner `i2c_base_addr->STAT = stat;`, just do that directly instead of calling this function below.

malto101 replied:
Got it.
I will update it!

glneo commented on drivers/i2c/i2c_omap.c:
```diff
@@ -0,0 +1,815 @@
+#define I2C_OMAP_WE_NACK_WE BIT(1)  /* No acknowledgment wakeup */
+#define I2C_OMAP_WE_AL_WE   BIT(0)  /* Arbitration lost wakeup */
+
+#define I2C_OMAP_WE_ALL                                                                            \
```
This define is only used once, just use the content in the one place it is used.

Speaking on that one place, do we know if it is needed? The wakeup support is probably not functional right now, you could just drop this an all the `I2C_OMAP_WE*` definitions in that case.

malto101 replied:
I have removed it!

glneo commented on drivers/i2c/i2c_omap.c:
```diff
@@ -0,0 +1,815 @@
+	uint32_t reg = i2c_base_addr->SYSTEST;
+
+	if (state) {
+		reg |= I2C_OMAP_SYSTEST_SCL_O;
```
You might be able to just avoid the temp variable `reg`, directly set/unset the bit, for instance here:

`i2c_base_addr->SYSTEST |= I2C_OMAP_SYSTEST_SCL_O;`

Same above in `i2c_omap_set_sda()` and `i2c_omap_resize_fifo()`

malto101 commented on drivers/i2c/i2c_omap.c:
```diff
@@ -0,0 +1,815 @@
+	/* Store the message in the data structure */
+	data->current_msg = *msg;
+	/* Insert a memory barrier to ensure proper instruction ordering */
+	compiler_barrier();
```
I had initially removed it prior to MMIO addition and it caused a lot of random result.

Edit: Just test it without it, and it worked.
Removed it.

glneo commented on drivers/i2c/i2c_omap.c:
```diff
@@ -0,0 +1,765 @@
+	i2c_omap_regs_t *i2c_base_addr = DEV_I2C_BASE(dev);
+
+	if (data->receiver) {
+		i2c_base_addr->BUF |= ((size) << 8) | I2C_OMAP_BUF_RXFIF_CLR;
```
You only seem to be OR'ing into ->BUF, how do these bits get cleared? I think you are missing some masking.

malto101 replied:
The mask was Rx/Tx DMA channel Enable. As we aren't using it as of now. I didn't mask it out.

glneo commented on drivers/i2c/i2c_omap.c:
```diff
@@ -0,0 +1,765 @@
+			data->cmd_err |= -EAGAIN;
+			return -EAGAIN;
+		}
+		data->cmd_err = err;
```
Having both `cmd_err` and `err` is confusing, sometimes we return one, other times we only add error bits to one, we even assign between them here. How about removing `err`. Everywhere we set it then return, could you just return it directly? Or set it in `data->cmd_err`. The only return code we ever check is for `-EAGAIN`, so why `return err;` below?

Also this function is already called in a loop that waits until stat is set, why nested loops at all? Could you combine the loops by simply removing the do-while loop from this function?

malto101 replied:
Ack
I will change it.

malto101 replied:
Removed the err variable and nested loop

glneo replied:
You removed the wrong loop, we want to keep the outer loop as it is the one that waits for events when polling. The inner loop simply tried to process all waiting events, but it doesn't itself wait once all events are cleared. Now we only wait once before trying to process all the events. If events are not queued up or fast enough now the transaction will fail. Idea was to still wait on status, but each call to `i2c_omap_transfer_message_ll()` would handle one event. That way the error handling could be simplified, it would just return any error directly.

malto101 replied:
My bad. I totally missed the part.

malto101 replied:
Hello @glneo .If the loop is removed from i2c_omap_transfer_message_ll, it would break the functionality needed for read-after-write, burst read, and burst write operations. The outer loop was essential because it waits for events during polling, allowing these transactions to complete successfully. The inner loop, by contrast, helps in processing multiple R/W I2C operations within a single transactions.

glneo replied:
Currently the outer-loop polls until at least one bit is set in `STAT`, then the inner-loop will loop until all bits set in `STAT` have been processed and cleared. After that the inner-loop exits and the outer-loop goes back to waiting for more `STAT` bits to come in. There is no need for the inner-loop as `i2c_omap_transfer_message_ll` could simply process one `STAT` bit per call, the outer-loop would then see some other `STAT` bits are still set and go right back to call `i2c_omap_transfer_message_ll` again, nothing is lost by removing the inner-loop.

What we gain by removing it is easier to understand flow control and error handling. Plus better timeout handling. Simply loop until the whole transfer is complete or we go too long without a `STAT` bit being set (timeout). This also makes the future IRQ support easier too, simply call the inner-function for each IRQ (which happens any time a `STAT` bit gets set).

malto101 replied:
I have removed the loop from `i2c_omap_transfer_message_ll`

malto101 replied:
I will be closing this.

nmenon commented on drivers/i2c/i2c_omap.c:
```diff
@@ -0,0 +1,762 @@
+			break;
+		}
+		result = -EAGAIN;
+	}
```
Something like this perhaps?
```
	for (uint8_t retries = 0; retries < 5, i2c_base_addr->STAT; retries++) 
                  add_some_delay();
        if (retries == 5) {
                   print error and exit
       }
```

dnltz reviewed (dismissed):
I finally had some time to test your driver. Please add a device-tree node for the BeagleAI and enable `CONFIG_I2C`. The current actions don't build this driver and therefore we don't see current build issues...

malto101 commented on drivers/i2c/i2c_omap.c:
```diff
@@ -0,0 +1,745 @@
+	}
+	if (stat & I2C_OMAP_STAT_NACK) {
+		data->cmd_err = -I2C_OMAP_STAT_NACK;
+		i2c_omap_ack_stat(dev, I2C_OMAP_STAT_NACK);
```
Updated!

malto101 commented on drivers/i2c/i2c_omap.c:
```diff
@@ -0,0 +1,745 @@
+	static void i2c_omap_##inst##_init(const struct device *dev)                               \
+	{                                                                                          \
+		IRQ_CONNECT(DT_INST_IRQN(inst), DT_INST_IRQ(inst, priority), i2c_omap_isr,         \
+			    DEVICE_DT_INST_GET(inst), DT_INST_IRQ(inst, sense));                   \
```
This was my first time messing with zephyr, so I may have left it there
It was supposed to be 
```
	static void i2c_omap_##inst##_init(const struct device *dev)                               \
	{                                                                                          \
		DEVICE_MMIO_NAMED_MAP(dev, base, K_MEM_CACHE_NONE);		\
		IRQ_CONNECT(DT_INST_IRQN(inst), DT_INST_IRQ(inst, priority),	\
			    i2c_omap_isr, DEVICE_DT_INST_GET(inst), 0);                   \
                                                                                                   \
		irq_enable(DT_INST_IRQN(inst));                                                    \
	};
```
Sorry for the trouble

malto101 committed:
boards: beagle: Enable I2C6 on BeagleBone AI64 board

Provide I2C Support to BeagleBone AI64 board.

Signed-off-by: Dhruv Menon <dhruvmenon1104@gmail.com>

malto101 commented:
Hello @dnltz, I’ve added the device tree node for the BeagleBone AI-64 and enabled CONFIG_I2C. However, since the current upstream support stores the code in ATCM instead of DDR, storage is quite limited.

dnltz commented:
> Hello @dnltz, I’ve added the device tree node for the BeagleBone AI-64 and enabled CONFIG_I2C. However, since the current upstream support stores the code in ATCM instead of DDR, storage is quite limited.

How much is currently used?

AM64x with Hello World reports the following to me which I think is totally fine.

```
Memory region         Used Size  Region Size  %age Used
           FLASH:          0 GB         0 GB
             RAM:       15332 B       192 KB      7.80%
             DDR:          16 B         4 KB      0.39%
        IDT_LIST:          0 GB        32 KB      0.00%
```

glneo commented on boards/beagle/beaglebone_ai64/beaglebone_ai64_j721e_main_r5f0_0.dts:
```diff
@@ -19,6 +19,7 @@
 	chosen {
 		zephyr,sram = &atcm;
 		zephyr,console = &uart2;
+		zephyr,shell-uart = &uart2;
```
This change is unrelated to this series (it seem correct, just belongs in another commit).

malto101 replied:
I will drop it.

glneo commented on dts/arm/ti/j721e_main_r5.dtsi:
```diff
@@ -55,6 +55,17 @@
 		status = "okay";
 	};
 
+	i2c6: i2c@2060000 {
```
Could you add all the MAIN domain I2C ports while here? I know I2C6 is pinned out in an easy way on Beaglebone-AI64 and the one you tested with, but this is the SoC DTSI file and should be complete. Should be easy, all the addresses and IRQs are numbers sequential.

malto101 replied:
No problem. I should've thought about it.
Will update it.

malto101 commented:
> How much is currently used?
> 
> AM64x with Hello World reports the following to me which I think is totally fine.
> 
> ```
> Memory region         Used Size  Region Size  %age Used
>            FLASH:          0 GB         0 GB
>              RAM:       15332 B       192 KB      7.80%
>              DDR:          16 B         4 KB      0.39%
>         IDT_LIST:          0 GB        32 KB      0.00%
> ```

I have attached the Hello world report below
```
[145/145] Linking C executable zephyr/zephyr.elf
Memory region         Used Size  Region Size  %age Used
           FLASH:          0 GB         0 GB
             RAM:       20384 B        32 KB     62.21%
            ATCM:          0 GB        32 KB      0.00%
            BTCM:          0 GB        32 KB      0.00%
       RSC_TABLE:          16 B         1 MB      0.00%
            DRAM:          0 GB        14 MB      0.00%
        IDT_LIST:          0 GB        32 KB      0.00%
```

DDR support adds an additional 14 MB. 
```
             RAM:       21164 B        14 MB      0.14%
```

glneo commented on drivers/i2c/i2c_omap.c:
```diff
@@ -0,0 +1,748 @@
+		stat &= ~(I2C_OMAP_STAT_RDR | I2C_OMAP_STAT_RRDY);
+	}
+	if (stat & I2C_OMAP_STAT_NACK) {
+		data->cmd_err = -I2C_OMAP_STAT_NACK;
```
Here you are assigning `-I2C_OMAP_STAT_NACK`, but below you check for `data->cmd_err & I2C_OMAP_STAT_NACK`. Should this line be `data->cmd_err |= I2C_OMAP_STAT_NACK;`?

Same below for `I2C_OMAP_STAT_AL`, but the issue would be you have defined your return code `RETRY` to be the same bit value.

Better question I now have is why do we need `cmd_err` at all? Why not just return the error directly from this function instead of storing is and fetching it out later in the calling function?

teburd commented on drivers/i2c/i2c_omap.c:
```diff
@@ -0,0 +1,746 @@
+{
+	uint32_t speed_cfg;
+	i2c_omap_regs_t *i2c_base_addr = DEV_I2C_BASE(dev);
+
```
There should be a semaphore used here as a lock to order access by threads

malto101 replied:
I have updated it.

teburd reviewed (changes requested):
Any of the i2c API functions should be protected by a semaphore to ensure a FIFO ordering of concurrent users of the i2c controller.

Imagine two sensors on the same bus for example attempting to read in parallel. The driver should ensure ordering of access.

While a k_mutex may be the ideal option, there's a code size cost and priority boosting has other downsides so the preferred primitive, as with SPI, is a semaphore.

Additionally.. the register map ideally would be done using #define's and sys_write/sys_read rather than a struct mapped to the base address to better match Zephyr driver style, but I won't block on this.

glneo commented on drivers/i2c/i2c_omap.c:
```diff
@@ -0,0 +1,710 @@
+	}
+
+	if (stat & I2C_OMAP_STAT_ROVR) {
+		result |= I2C_OMAP_STAT_ROVR;
```
You could just return `I2C_OMAP_STAT_ROVR` directly, no need to add to `result` just to return that in the next line.

malto101 replied:
Removed it

malto101 replied:
Sorry for that.
I have updated it now.

glneo commented on drivers/i2c/i2c_omap.c:
```diff
@@ -0,0 +1,710 @@
+ * @param addr Target address.
+ * @return 0 on success, negative error code on failure.
+ */
+static int __maybe_unused i2c_omap_transfer_polling(const struct device *dev, struct i2c_msg msgs[],
```
Why `__maybe_unused`? Looks like it will always be used (at least until interrupt support is added, at that point if it is selectable then you can add this back).

malto101 replied:
Removed it for now
Maybe I could add a property to let us choose between polling and interrupt, once I add support for it

malto101 commented:
Hello @vaishnavachath @teburd,
 I’ve incorporated the suggested changes and believe the PR is now ready. If everything looks good on your end, could you kindly approve it?

vaishnavachath commented on boards/beagle/beaglebone_ai64/beaglebone_ai64_j721e_main_r5f0_0.yaml:
```diff
@@ -14,4 +14,5 @@ toolchain:
 supported:
   - uart
+  - i2c
 vendor: beagle
```
@malto101 Can you please update in https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/beagle/beaglebone_ai64/doc/index.rst#supported-features also?

malto101 replied:
I Have updated the docs to show the same.

pdgendt commented on drivers/i2c/i2c_omap.c:
```diff
@@ -0,0 +1,708 @@
+	return i2c_omap_transfer_main(dev, msgs, num_msgs, true, addr);
+}
+
+static const struct i2c_driver_api i2c_omap_api = {
```
Device driver APIs need to be wrapped using `DEVICE_API`.

```suggestion
static DEVICE_API(i2c, i2c_omap_api) = {
```

malto101 replied:
Hello @pdgendt , thank you for the review, Most of the I2C controller implementation have implemented in the former way,
Could you give me a reference for wrapping it with `DEVICE_API`

pdgendt replied:
This is a recent addition to the [device driver model](https://docs.zephyrproject.org/latest/kernel/drivers/index.html#subsystems-and-api-structures), since https://github.com/zephyrproject-rtos/zephyr/pull/71773 has been merged.

https://github.com/zephyrproject-rtos/zephyr/pull/82207 will be merged soon for the existing I2C+I3C drivers.

malto101 replied:
Thanks for the quick update,
I have updated it!

pdgendt commented on boards/beagle/beaglebone_ai64/beaglebone_ai64_j721e_main_r5f0_0_defconfig:
```diff
@@ -13,3 +13,6 @@ CONFIG_UART_INTERRUPT_DRIVEN=y
 CONFIG_CONSOLE=y
 CONFIG_UART_CONSOLE=y
+
+# Enable I2C
+CONFIG_I2C=y
```
This shouldn't go in the `defconfig`.

malto101 replied:
I had recently append this
it is with relation with https://github.com/zephyrproject-rtos/zephyr/pull/80113#pullrequestreview-2416421386 comment

pdgendt replied:
I think you should add `tests/drivers/i2c/i2c_api/boards/beaglebone_ai64_...overlay` to make it build there.

malto101 replied:
Removed CONFIG_I2C

malto101 replied:
Hello @pdgendt do I need to add the above?, as it won't be able to compile due to RAM overflow until DDR support comes upstream.

pdgendt replied:
The `_defconfig` file is reserved for options that a board minimally needs to be able to boot. You should make sure that it's get build/tested in the tests/samples that only depend on I2C.

malto101 replied:
Ahh, Thanks for clarifying it,
I will be closing this conversation then.

pdgendt commented on drivers/i2c/Kconfig.omap:
```diff
@@ -0,0 +1,18 @@
+	default y
+	depends on DT_HAS_TI_OMAP_I2C_ENABLED
+	help
+		Enable the I2C driver for TI OMAP SoCs.
```
Help text should be indented with a tab and 2 spaces.
```suggestion
	  Enable the I2C driver for TI OMAP SoCs.
```

malto101 replied:
Updated it.

pdgendt commented on drivers/i2c/Kconfig.omap:
```diff
@@ -0,0 +1,18 @@
+	depends on I2C_OMAP
+	select I2C_BITBANG
+	help
+		Enable OMAP I2C driver bus recovery support via bitbanging.
```
```suggestion
	  Enable OMAP I2C driver bus recovery support via bitbanging.
```

malto101 replied:
updated it.

malto101 commented:
hello @pdgendt, If everything looks fine, Could you approve the PR?

malto101 commented:
There was merge conflict with CMakeList.txt and Config, fixed it and pushed it
cc: @teburd @pdgendt @vaishnavachath
